program-runtime/src/loaded_programs.rs

```diff
@@ -914,6 +952,39 @@ impl<FG: ForkGraph> LoadedPrograms<FG> {
         self.unload_program_entries(sorted_candidates.iter().take(num_to_unload));
     }
 
+    /// Evicts programs using 2's random selection, choosing the least used program out of the two entries.
+    /// The eviction is performed enough number of times to reduce the cache usage to the given percentage.
+    pub fn evict_using_2s_random_selection(&mut self, shrink_to: PercentageInteger, now: Slot) {
+        let mut candidates = self.get_flattened_entries();
```

Comment:
could switch `self.entries` from `HashMap` to `IndexMap`. then the draw is just two random samplings from `0..self.entries.len()`. no allocation necessary

Reply:
Changing to `IndexMap` will impact other parts of the code in the cache, not just eviction logic. If we agree to do it, maybe best to do it in a separate PR (in case we want to backport this). Thoughts?

Reply:
Each entry in `entries` is a `Vec<>` of cached programs (since a given program ID can potentially have different versions of the program loaded simultaneously in cache corresponding to different slot/fork). So we need to flatten the two dimensional list to actually compute the total cache usage, and count how much to evict. I don't think we can avoid allocation even if we were to use IndexMap here.

Reply:
This sounds like a case for a BTreeMap? With keys being (program_key, slot) or something

Reply:
(I think that github is having issues with PR, my comments keep appearing and disappearing, but since this disappeared again)

I think this sounds like we should use a BTreeMap with keys (pubkey, slot)? So we can key ranges and still have a total length. It's a tiny map so perf wise it's almost certainly better than juggling a two level hash table + vec.

Reply:
We do a reverse lookup for entries of a program during `extract()`. This helps find the most latest version of the program applicable to the given slot. That was one of the main reason/advantage of doing two level table.

Maybe this is not a big perf hit if we were to use BTreeMap. But, I think we should analyze it more carefully.

Reply:
BTreeMap::range supports reverse iteration, I don't think it's going to be any slower than what we're currently doing?

Reply:
We can/should certainly explore it. I believe it'll intersect with the other PR that does cooperative loading of programs.

Filed this issue to track it: https://github.com/solana-labs/solana/issues/34442